ada/src/stemmer.ads

```diff
@@ -0,0 +1,219 @@
+   --  A modular value is used because shift and logical arithmetic is necessary.
+   type Utf8_Type is mod 2**32;
+
+   --  Index of the Grouping_Array.  The index comes from the 32-bit character value
+   --  minus a starting offset.  We don't expect large tables and we check against
+   --  a maximum value.
+   subtype Grouping_Index is Utf8_Type range 0 .. 16384;
+
+   type Grouping_Array is array (Grouping_Index range <>) of Boolean with Pack;
+
+   subtype Among_Index is Natural range 0 .. 65535;
+   subtype Among_Start_Index is Among_Index range 1 .. Among_Index'Last;
+   subtype Operation_Index is Natural range 0 .. 65535;
+   subtype Result_Index is Integer range -1 .. WORD_MAX_LENGTH - 1;
+   subtype Char_Index is Result_Index range 0 .. Result_Index'Last;
+
+   type Among_Type is record
+      First       : Among_Start_Index;
+      Last        : Among_Index;
+      Substring_I : Integer;
+      Result      : Integer;
+      Operation   : Operation_Index;
+   end record;
+
+   type Among_Array_Type is array (Natural range <>) of Among_Type;
+
+   function Eq_S (Context : in Context_Type'Class;
+                  S       : in String) return Char_Index with
+     Global => null,
+     Pre => S'Length > 0,
+     Post => Eq_S'Result = 0 or Eq_S'Result = S'Length;
+
+   function Eq_S_Backward (Context : in Context_Type'Class;
+                           S       : in String) return Char_Index with
+     Global => null,
+     Pre => S'Length > 0,
+     Post => Eq_S_Backward'Result = 0 or Eq_S_Backward'Result = S'Length;
+
+   procedure Find_Among (Context : in out Context_Type'Class;
+                         Amongs  : in Among_Array_Type;
+                         Pattern : in String;
+                         Execute : access procedure
+                           (Ctx       : in out Context_Type'Class;
+                            Operation : in Operation_Index;
+                            Status    : out Boolean);
+                         Result  : out Integer) with
+     Global => null,
+     Pre => Pattern'Length > 0 and Amongs'Length > 0;
+
+   procedure Find_Among_Backward (Context : in out Context_Type'Class;
+                                  Amongs  : in Among_Array_Type;
+                                  Pattern : in String;
+                                  Execute : access procedure
+                                    (Ctx       : in out Context_Type'Class;
+                                     Operation : in Operation_Index;
+                                     Status    : out Boolean);
+                                  Result  : out Integer) with
+     Global => null,
+     Pre => Pattern'Length > 0 and Amongs'Length > 0;
+
+   function Skip_Utf8 (Context : in Context_Type'Class) return Result_Index with
+     Global => null;
+
+   function Skip_Utf8 (Context : in Context_Type'Class;
+                       N       : in Positive) return Result_Index with
```

Comment:
There's a minor issue here.

In the [snowball manual](https://snowballstem.org/compiler/snowman.html), `hop AE` is documented as raising signal `f` if `AE` is negative, but currently Ada fails on `hop -1` with:

```
Execution of ./bin/stemwords terminated by unhandled exception
raised CONSTRAINT_ERROR : stemmer-yiddish.adb:1045 range check failed
```

This was flagged up by testing a merge with current git master as the recently merged Yiddish algorithm currently relies on `hop -1` currently being incorrectly implemented in existing backends to move the cursor back.  We're working on fixing this algorithm and the existing backends, so Ada should handle this case by raising `f` as documented rather than following how the other backends currently behave.

So either parameter `N` here and in the backwards variant should accept signed values, with its sign checked inside, or there should be a sign check before the call (which I suppose may make it easier for the compiler to optimise the check away in the common case where `skip` is used with a constant argument).

Reply:
The Ada generator produces:

```Ada
C := Skip_Utf8_Backward (Z, -5);
```

where the C generator produces:

```
int ret = skip_utf8(z->p, z->c, z->lb, z->l, - -5);
```

Indeed the `skip_utf8` should accept only positive values. It does not make sense to pass negative values to that function. The Ada compiler detected the error at compilation time:

```
stemmer-yiddish.adb:489:26: warning: value not in range of type "Standard.Positive"
stemmer-yiddish.adb:489:26: warning: "Constraint_Error" will be raised at run time
```

And then it fails at runtime as expected.

I'll have a look at the Ada code generator to check why the negation is dropped.